### ISSUE TITLE
npmhhome can not connect to npm

### ISSUE THREAD

**Issue by vvstubbs** [billettg/npmhome#10]
This afternoon I noticed that npmhome was not connecting, which was strange as nothing had changed, but I checked all the usual suspects, network connectivity, firewalls, docker, the container it's self and restarting and or recreating all of the a fore mentioned as required. The npmhome page itself starts, but in the upper right hand corner a icon of a plug with a small x indicating not connected. I checked the config.json, and all is good. Now as watchtower is on the host that npmhome is residing I deleted the container and the image, re-pulled the image, and recreated the container, but to no availe still not connecting. Ideas?

**Comment by billettg** [comment]
The cross indicates that the server can't connect to the NPM instance. I would need to see your docker compose/run command, config.json (omit any sensitive info), evidence of which port NPM webui/api is listening on (default 81), debug console output in browser and container log. Normally this is either misconfigured config, port is blocked, or NPM not listening.

**Comment by vvstubbs** [comment]
okay:
config.json:
{
    "server": {
        "host" : "192.168.0.183",
        "port" : 81,
        "email" : "admin@xxxxx.net",
       "password" : "Pxxxxxxxxxxxxxx!"
    },
    "appearance": {
        "showheader": true,
        "showfooter": true,
        "darkmode": true
    }
}

docker-compose.yml:

services:
    npmhome:
        image: billettg/npmhome
        container_name: npmhome
        volumes:
            - ./config.json:/app/config.json
        ports:
            - 1234:1234
        restart: unless-stopped

**Comment by billettg** [comment]
I've pushed a new build out just now with a fix which is available on the docker hub, please try it and let me know. I mistakenly forgot to whitelist my own config.json so it was added to the build and may be overwriting your own config.

Just so you are aware I am making major changes to this app as it's undergoing development, and will also be changing from mainly client to server side (node) architecture, so there may be some issues along the way.

**Comment by vvstubbs** [comment]
sorry forgot it doesn't keep the formating

**Comment by billettg** [comment]
Sorry deleted earlier comment as I discovered the issue was my side.

**Comment by vvstubbs** [comment]
ummm, well different, it just comes up with a completely white screen, with just npmhome in the left corner and the github... in the middle, but no line in fact nothing else at all

**Comment by billettg** [comment]
Ah, sorry about that, I figured it out and will deploy a fix shortly. I've recently moved over to using a more server side architecture so making quite a lot of changes recently.

**Comment by vvstubbs** [comment]
yes, okay I did that, and the screen starts up but still can not connect. Probably the changes you doing, don't rush, as will all early development there can be breaking changes

**Comment by billettg** [comment]
That's odd as it works for me now, it was just an incorrect volume mapping as files are now served from the /public folder, which is where npmhome expects the config.json to be. What does the debug console for the browser show? Make sure you clear the browser cache (not just CTRL+F5). If you get stuck I'll help more tomorrow.

I'll probably label this as under development to make it clear that there may be breaking changes.

**Comment by vvstubbs** [comment]
Well here is some info from the browser debugger:

->Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://192.168.0.183:81/api/tokens. (Reason: CORS request did not succeed). Status code: (null).
->Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://192.168.0.183:81/api/nginx/proxy-hosts. (Reason: CORS request did not succeed). Status code: (null).
->Error: TypeError: NetworkError when attempting to fetch resource. index.js:71:13
    getProxyHosts http://192.168.0.183:1234/index.js:71
->Error: Promised response from onMessage listener went out of scope init.js:79:27
->Source map error: NetworkError when attempting to fetch resource.
Resource URL: moz-extension://9f7d7ff6-3dd9-4cc4-88d4-1c416af5e4d8/browser-polyfill.min.js
Source Map URL: browser-polyfill.min.js.map

**Comment by billettg** [comment]
Ah it's blocked by CORS policy and it's always causing issues with cross origin requests. I could probably fix this by routing the connection via the cors-anywhere transparent proxy I built into the image which I had to do for favicon searches. I'll add it to my todo list. New build coming soon which will be server orientated and requests to NPM coming from the server not the client.

**Comment by billettg** [comment]
Out of interest what happens if you enter the API address in your browser? It's odd because I access the NPM API in code and via curl or web browser fine with the bearer token.

**Comment by vvstubbs** [comment]
what is the url for the api?

**Comment by billettg** [comment]
The one shown in your log, http://192.168.0.183:81/api/tokens

**Comment by vvstubbs** [comment]
![Screenshot_2024-03-28_15-18-03](https://github.com/billettg/npmhome/assets/49411467/a22318ee-74ef-415c-a5b5-62aa80661aba)

**Comment by vvstubbs** [comment]
It might be that I'm not directly on 192.168.0.183, it's headless and has no GUI , so it from the browser on my workstation

**Comment by billettg** [comment]
Looks more like a problem accessing NPM in general. When I visit that URL remotely in my browser I see:

{
  "error": {
    "code": 400,
    "message": "Internal Error"
  }
}

I presume NPM is running and you can access http://192.168.0.183:81 ?

**Comment by vvstubbs** [comment]
I made it work!, I moved NPM so it was on the same host as npmhome, on the same docker network. It was fairly easy as all my docker apps use one of two central databases, NPM need MySQL so it point to the host that has a bare metal mariadb running, so it was just move the folders over, and it works, nice icons by the way...
When you get the multi arch going I put them both back on the PI4, Strange though all the correct ports where open between the two machines, and I have never had a problem with different containers on different machines, even if the base network address ever give issues I then just use the tailscale address the each machine has. (each machine is multi homed, and all the main hosts have 4-6 NIC's the PI4 is the only thing close to normal, but even it has a tailscale address as well as bare metal IP.

**Comment by vvstubbs** [comment]
So Sorry I wasted your time, if I'd just use the software the way it was intended, (on the same machine, in the same docker instance). By the way Is one of the "server" changes for it to ask for a login? cool...

**Comment by billettg** [comment]
Good to hear you got it sorted. CORS can be a real PITA even blocking requests from the same subnet or even same IP sometimes which is why there are solutions like cors-anywhere to respond with the required allow headers. As I said in a new build requests will actually come from the server rather than the client and routed via the CORS proxy transparently so shouldn't see any more CORS issues like this. In fact, everything will be handled server side only sending necessary data to the client for rendering so the whole app will be more secure and robust.

Glad you like the icons. Took me a while to get that working! Had to do a 3 stage request of the base url, then scouring html for icon links and finally walkxcodes dashboard icons on the web, all routed via the CORS proxy again, and then cached on the client. Phew.
